File: import_products.py
import sqlite3
import re
import os

db_path = r".dart_tool\sqflite_common_ffi\databases\nextgen_billing.db"
txt_path = "raw_products_list.txt"

if not os.path.exists(db_path):
    print(f"Error: {db_path} not found.")
    exit(1)

conn = sqlite3.connect(db_path)
cursor = conn.cursor()

with open(txt_path, 'r', encoding='utf-8') as f:
    lines = f.readlines()

success_count = 0
error_count = 0

for line in lines:
    line = line.strip()
    if not line:
        continue

    name = line
    price = 0.0
    category_id = 7 # Default to others

    # Regex to extract price if it ends with XXRS or XX RS
    match = re.search(r'([\d.]+)\s*RS', line, re.IGNORECASE)
    if match:
        price = float(match.group(1))

    # Category assignment
    lower = line.lower()
    if any(k in lower for k in ['milk', 'tea', 'coffee', 'curd', 'butter', 'cheese']):
        category_id = 1
    elif any(k in lower for k in ['oil', 'ghee']):
        category_id = 2
    elif any(k in lower for k in ['biscuit', 'cake', 'cookie', 'chocolate', 'snack', 'chips', 'rusk', 'sweet']):
        category_id = 3
    elif any(k in lower for k in ['shampoo', 'soap', 'paste', 'brush', 'powder', 'cream', 'liquid', 'perfume', 'wash']):
        category_id = 4
    elif any(k in lower for k in ['rice', 'atta', 'maida', 'dal', 'parupu', 'gothumai', 'flour', 'kurunai', 'ulundhu']):
        category_id = 5
    elif any(k in lower for k in ['masala', 'thool', 'spice', 'pepper', 'salt', 'sugar', 'podi']):
        category_id = 6

    try:
        cursor.execute('''
            INSERT INTO products (
                name, selling_price, category_id, current_stock, unit, unit_value, gst_percentage, purchase_price, wholesale_price, hotel_price, min_stock, barcode, sku
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (name, price, category_id, 100.0, 'Piece', 1.0, 18.0, 0.0, 0.0, 0.0, 0.0, '', ''))
        success_count += 1
    except sqlite3.Error as e:
        error_count += 1
        print(f"Failed to insert {name}: {e}")

conn.commit()
conn.close()

print(f"Import complete! Successfully inserted: {success_count}, Errors: {error_count}")

File: import_products.dart
import 'dart:io';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:path/path.dart';

void main() async {
  // Initialize FFI
  sqfliteFfiInit();
  var databaseFactory = databaseFactoryFfi;
  
  String dbPath = join(await databaseFactory.getDatabasesPath(), 'nextgen_billing.db');
  print('Connecting to database at: $dbPath');
  
  var db = await databaseFactory.openDatabase(dbPath);

  File file = File('raw_products_list.txt');
  if (!await file.exists()) {
    print('Error: raw_products_list.txt not found.');
    exit(1);
  }
  
  List<String> lines = await file.readAsLines();
  print('Loaded ${lines.length} products. Starting import...');

  int successCount = 0;
  int errorCount = 0;

  for (String line in lines) {
    line = line.trim();
    if (line.isEmpty) continue;

    String name = line;
    double price = 0.0;
    int categoryId = 7; // Default to Fresh Vegetables & Greens/Others

    // Extract price if it ends with XXRS or XX RS
    RegExp regExp = RegExp(r'([\d.]+)\s*RS', caseSensitive: false);
    var match = regExp.firstMatch(line);
    if (match != null) {
      price = double.tryParse(match.group(1) ?? '0') ?? 0.0;
    }

    // Category assignment based on keywords
    String lower = line.toLowerCase();
    if (lower.contains('milk') || lower.contains('tea') || lower.contains('coffee') || lower.contains('curd') || lower.contains('butter') || lower.contains('cheese')) {
      categoryId = 1; // Dairy & Beverages
    } else if (lower.contains('oil') || lower.contains('ghee')) {
      categoryId = 2; // Oils & Ghee
    } else if (lower.contains('biscuit') || lower.contains('cake') || lower.contains('cookie') || lower.contains('chocolate') || lower.contains('snack') || lower.contains('chips') || lower.contains('rusk') || lower.contains('sweet')) {
      categoryId = 3; // Snacks & Bakery
    } else if (lower.contains('shampoo') || lower.contains('soap') || lower.contains('paste') || lower.contains('brush') || lower.contains('powder') || lower.contains('cream') || lower.contains('liquid') || lower.contains('perfume') || lower.contains('wash')) {
      categoryId = 4; // Personal Care
    } else if (lower.contains('rice') || lower.contains('atta') || lower.contains('maida') || lower.contains('dal') || lower.contains('parupu') || lower.contains('gothumai') || lower.contains('flour') || lower.contains('kurunai') || lower.contains('ulundhu')) {
      categoryId = 5; // Grains
    } else if (lower.contains('masala') || lower.contains('thool') || lower.contains('spice') || lower.contains('pepper') || lower.contains('salt') || lower.contains('sugar') || lower.contains('podi')) {
      categoryId = 6; // Spices & Masalas
    }

    try {
      await db.insert('products', {
        'name': name,
        'selling_price': price,
        'category_id': categoryId,
        'current_stock': 100.0,
        'unit': 'Piece',
        'unit_value': 1.0,
        'gst_percentage': 18.0,
      }, conflictAlgorithm: ConflictAlgorithm.ignore);
      successCount++;
    } catch (e) {
      errorCount++;
      print('Failed to insert $name: $e');
    }
  }

  print('Import complete! Successfully inserted: $successCount, Errors: $errorCount');
  await db.close();
  exit(0);
}
